engine/graphics/src/opengl/graphics_opengl.cpp

```diff
@@ -1739,60 +1801,140 @@ static uintptr_t GetExtProcAddress(const char* name, const char* extension_name,
+        //       so we need to store a copy of the texture array source
+        //       since the programs that use this variant will patch in the
+        //       maximum number of pages supported before creating the 
+        //       shader program.
+        if (ddf->m_VariantTextureArray)
+        {
+            Shader* shader = new Shader();
+            shader->m_VariantTextureArrayData     = (char*) malloc(ddf->m_Source.m_Count);
+            shader->m_VariantTextureArrayDataSize = ddf->m_Source.m_Count;
+            memcpy(shader->m_VariantTextureArrayData, ddf->m_Source.m_Data, ddf->m_Source.m_Count);
+            return shader;
+        }
+        else
+        {
+            GLuint shader_id = DoCreateShader(type, ddf->m_Source.m_Data, ddf->m_Source.m_Count);
+            if (!shader_id)
+            {
+                return 0;
+            }
+            Shader* shader = new Shader();
+            shader->m_Id = shader_id;
+            return shader;
+        }
     }
 
     static HVertexProgram OpenGLNewVertexProgram(HContext context, ShaderDesc::Shader* ddf)
     {
-        assert(ddf);
-        return CreateShader(GL_VERTEX_SHADER, ddf->m_Source.m_Data, ddf->m_Source.m_Count);
+        return (HVertexProgram) CreateShader(GL_VERTEX_SHADER, ddf);
     }
 
     static HFragmentProgram OpenGLNewFragmentProgram(HContext context, ShaderDesc::Shader* ddf)
     {
-        assert(ddf);
-        return CreateShader(GL_FRAGMENT_SHADER, ddf->m_Source.m_Data, ddf->m_Source.m_Count);
+        return (HFragmentProgram) CreateShader(GL_FRAGMENT_SHADER, ddf);
+    }
+
+    static GLuint CreateShaderVariantTextureArray(Shader* shader, GLenum type, uint8_t max_page_count)
+    {
+        assert(max_page_count > 0);
+
+        char* source_copy = (char*) malloc(shader->m_VariantTextureArrayDataSize);
+        memcpy(source_copy, shader->m_VariantTextureArrayData, shader->m_VariantTextureArrayDataSize);
+
+        // This is what shaderutil.java adds in bob when producing the variant, we need to replace val with our max value
+        const char* dm_array_define_str = "#define DM_MAX_PAGE_COUNT ###";
+
+        char* array_define_start = strstr(source_copy, dm_array_define_str);
+        assert(array_define_start);
+
+        char array_define_buf[32] = {0};
+
+        snprintf(array_define_buf, sizeof(array_define_buf), "#define DM_MAX_PAGE_COUNT %3d", max_page_count); // %3d == sizeof("###")
+        strncpy(array_define_start, array_define_buf, strlen(array_define_buf));
+
+        GLuint handle = DoCreateShader(type, source_copy, shader->m_VariantTextureArrayDataSize);
+        free(source_copy);
+
+        return handle;
     }
 
-    static HProgram OpenGLNewProgram(HContext context, HVertexProgram vertex_program, HFragmentProgram fragment_program)
+    static HProgram OpenGLNewProgram(HContext context, const ProgramCreationParams& params)
     {
         IncreaseModificationVersion(context);
 
+        Program* program = new Program();
```

Comment:
A program was previously just a GL id but now it's an actual pointer since it contains more data